src/mint.rs

```diff
@@ -0,0 +1,190 @@
+use codec::{Decode, Encode};
+use runtime_primitives::traits::{SimpleArithmetic, Zero};
+use srml_support::ensure;
+
+#[derive(Encode, Decode, Copy, Clone, Debug, Eq, PartialEq)]
+pub enum AdjustCapacityBy<Balance> {
+    // Set capacity of mint to specific value
+    Setting(Balance),
+    // Add to the capacity of the mint
+    Adding(Balance),
+    // Reduce capacity of the mint
+    Reducing(Balance),
+}
+
+#[derive(Encode, Decode, Copy, Clone, Debug, Eq, PartialEq)]
+pub struct AdjustOnInterval<Balance, BlockNumber> {
+    pub block_interval: BlockNumber,
+    pub adjustment_type: AdjustCapacityBy<Balance>,
+}
+
+#[derive(Encode, Decode, Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Adjustment<Balance, BlockNumber> {
+    // First adjustment will be after AdjustOnInterval.block_interval
+    Interval(AdjustOnInterval<Balance, BlockNumber>),
+    // First Adjustment will be at absolute blocknumber
+    IntervalAfterFirstAdjustmentAbsolute(AdjustOnInterval<Balance, BlockNumber>, BlockNumber),
+    // First Adjustment will be after a specified number of blocks
+    IntervalAfterFirstAdjustmentRelative(AdjustOnInterval<Balance, BlockNumber>, BlockNumber),
+}
+
+#[derive(Encode, Decode, Default)]
+// Note we don't use TokenMint<T: Trait> it breaks the Default derivation macro with error T doesn't impl Default
+// Which requires manually implementing Default trait.
+// We want Default trait on TokenMint so we can use it as value in StorageMap without needing to wrap it in an Option
+pub struct Mint<Balance, BlockNumber>
+where
+    Balance: Copy + SimpleArithmetic,
+    BlockNumber: Copy + SimpleArithmetic,
+{
+    capacity: Balance,
+
+    adjustment_on_interval: Option<AdjustOnInterval<Balance, BlockNumber>>,
+
+    // Whether there is an upcoming block where
+    // When this is not set, the mint is effectively paused.
+    adjust_capacity_at_block_number: Option<BlockNumber>,
+
+    created_at: BlockNumber,
+
+    total_minted: Balance,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum MintingError {
+    NotEnoughCapacity,
+    InvalidMint,
+    InvalidSourceMint,
+    InvalidDestinationMint,
+    NextAdjustmentInPast,
+}
+
+impl<Balance, BlockNumber> Mint<Balance, BlockNumber>
+where
+    Balance: Copy + SimpleArithmetic,
+    BlockNumber: Copy + SimpleArithmetic,
+{
+    pub fn new(
```

Comment:
I am not sure we want to force new mints to start at fixed adjustment offset from `now`. Perhaps caller wants to start with an arbitrary beginning delay, that was the entire rationale for actually representing the state this way if you recall. See https://github.com/Joystream/joystream/issues/106#issuecomment-524769707

"Yeah, that is very simple, but then it means we cannot independently control when the first payoff is, and how frequently payoffs are made. "

I think we must chose, either we want to support that flexibility, and thus change this, or just throw the idea away and go with a simpler representation to begin with.

Reply:
The variant `IntervalAfterFirstAdjustmentAbsolute(AdjustOnInterval<Balance, BlockNumber>, BlockNumber)` allows the for setting an arbitrary blocknumber when to start minting.